### PR TITLE
Add support for push notifications

### PR DESCRIPTION
**Pull request by AmarOk1412**
cf #225

+ Add SUBSCRIBE and UNSUBSCRIBE for DhtProxyServer
+ Add the ability to stop Listen via a timeout (and send a push
Notification for the client)
+ Add the ability to use gorush to send push notifications
+ add the ability to inject push notifications in opendht to call
correct callbacks
+ if a timeout occurs and no cancelListen was called, relaunch the
listen operation
+ Pass devicekey to enable push notifications for DhtProxyClient
+ Update sockaddr to fix the build on all platforms

Note: Update dhtnode

### CODE REVIEW

File: src/dht_proxy_client.cpp [hunk 1]
```diff
@@ -499,13 +533,33 @@ DhtProxyClient::cancelListen(const InfoHash&, size_t token)
...
                 return true;
+            } else {
+                // Just stop the request
+                if (listener.thread.joinable()) {
+                    // Close connection to stop listener?
+                    if (listener.req)
+                        restbed::Http::close(listener.req);
+                    listener.thread.join();
+                    listeners_.erase(it);
+                    lockListener_.unlock();
```
Comment (aberaud) [inline src/dht_proxy_client.cpp]: not necessary: mutex is unlocked on destruction ok lock_guard

File: tools/dhtnode.cpp [hunk 2]
```diff
@@ -101,13 +113,17 @@ void cmd_loop(std::shared_ptr<DhtRunner>& dht, dht_params& params)
 #if OPENDHT_PROXY_SERVER
```
Comment (aberaud) [inline tools/dhtnode.cpp]: This up to line 128 should not be part of the cmd_loop, should go to main, to be respected even in non-interactive modes. Maybe make proxies a global, of pass reference to cmd_loop ?

File: tools/dhtnode.cpp [hunk 3]
```diff
@@ -472,6 +510,15 @@ main(int argc, char **argv)
         }
 
         if (params.daemonize or params.service) {
+#if OPENDHT_PROXY_SERVER
+            if (params.proxyserver != 0) {
+                new DhtProxyServer(dht, params.proxyserver
```
Comment (aberaud) [inline tools/dhtnode.cpp]: See comment above (also true for proxyclient). Here, the object lifetime is not correctly handled, so the server is not shutdown properly on exit.

File: tools/dhtnode.cpp (hunk 2 above)
Comment (AmarOk1412) [reply tools/dhtnode.cpp]: I prefer the reference.

**Pull request merged**